File: conformance/tests/math-demo/math-demo.runar.zig
const runar = @import("runar");

pub const MathDemo = struct {
    pub const Contract = runar.StatefulSmartContract;

    value: i64 = 0,

    pub fn init(value: i64) MathDemo {
        return .{ .value = value };
    }

    pub fn divideBy(self: *MathDemo, divisor: i64) void {
        self.value = runar.safediv(self.value, divisor);
    }

    pub fn withdrawWithFee(self: *MathDemo, amount: i64, feeBps: i64) void {
        const fee = runar.percentOf(amount, feeBps);
        const total = amount + fee;
        runar.assert(total <= self.value);
        self.value = self.value - total;
    }

    pub fn clampValue(self: *MathDemo, lo: i64, hi: i64) void {
        self.value = runar.clamp(self.value, lo, hi);
    }

    pub fn normalize(self: *MathDemo) void {
        self.value = runar.sign(self.value);
    }

    pub fn exponentiate(self: *MathDemo, exp: i64) void {
        self.value = runar.pow(self.value, exp);
    }

    pub fn squareRoot(self: *MathDemo) void {
        self.value = runar.sqrt(self.value);
    }

    pub fn reduceGcd(self: *MathDemo, other: i64) void {
        self.value = runar.gcd(self.value, other);
    }

    pub fn scaleByRatio(self: *MathDemo, numerator: i64, denominator: i64) void {
        self.value = runar.mulDiv(self.value, numerator, denominator);
    }

    pub fn computeLog2(self: *MathDemo) void {
        self.value = runar.log2(self.value);
    }
};

File: conformance/tests/math-demo/math-demo.runar.go
//go:build ignore

package contract

import "runar"

type MathDemo struct {
	runar.StatefulSmartContract
	Value runar.Bigint
}

func (c *MathDemo) DivideBy(divisor runar.Bigint) {
	c.Value = runar.Safediv(c.Value, divisor)
}

func (c *MathDemo) WithdrawWithFee(amount, feeBps runar.Bigint) {
	fee := runar.PercentOf(amount, feeBps)
	total := amount + fee
	runar.Assert(total <= c.Value)
	c.Value = c.Value - total
}

func (c *MathDemo) ClampValue(lo, hi runar.Bigint) {
	c.Value = runar.Clamp(c.Value, lo, hi)
}

func (c *MathDemo) Normalize() {
	c.Value = runar.Sign(c.Value)
}

func (c *MathDemo) Exponentiate(exp runar.Bigint) {
	c.Value = runar.Pow(c.Value, exp)
}

func (c *MathDemo) SquareRoot() {
	c.Value = runar.Sqrt(c.Value)
}

func (c *MathDemo) ReduceGcd(other runar.Bigint) {
	c.Value = runar.Gcd(c.Value, other)
}

func (c *MathDemo) ScaleByRatio(numerator, denominator runar.Bigint) {
	c.Value = runar.MulDiv(c.Value, numerator, denominator)
}

func (c *MathDemo) ComputeLog2() {
	c.Value = runar.Log2(c.Value)
}
